pkg/filters/builder/requestadaptor/requestadaptor.go

```diff
@@ -248,6 +254,22 @@ func (ra *RequestAdaptor) Handle(ctx *context.Context) string {
 	req := ctx.GetInputRequest().(*httpprot.Request)
 	method, path := req.Method(), req.Path()
 
+	if ra.spec.Template != "" {
+		data, err := builder.PrepareBuilderData(ctx)
+		if err != nil {
+			logger.Warnf("PrepareBuilderData failed: %v", err)
+			return builder.ResultBuildErr
+		}
+
+		tempSpec := &Spec{}
+		if err = ra.Builder.Build(data, tempSpec); err != nil {
+			msgFmt := "RequestAdaptor(%s): failed to build adaptor info: %v"
+			logger.Warnf(msgFmt, ra.Name(), err)
+			return builder.ResultBuildErr
+		}
+		ra.adaptSpecWithTemplate(tempSpec)
```

Comment:
`Handle` will be called concurrently, modifying the spec result in race conditions.
Same issue for the `ResponseAdaptor`.

Reply:
There are 3 ways to fix the race condition.
- Deep copy the static spec. As the length of static spec is usually small, it's not that costly.
- Give a lock to the static spec, and modify the spec with reference. No deep copy, but the route with template become sequence, and it's costly.
- keep static spec mutually exclusive with template as RequestBuilder does. Lose flexible.

What's your opinion?

Reply:
none of the three is good enough, please just use simple if/else, if template has the field then use template, otherwise, use the spec.

Reply:
I choose first method. If not, it may looks like
```golang
	if tempSpec.Method != "" && ra.spec.Method != method {
		ctx.AddTag(stringtool.Cat("requestAdaptor: method ", method, " adapted to ", tempSpec.Method))
		req.SetMethod(tempSpec.Method)
	} else if tempSpec.Method == "" && ra.spec.Method != "" && ra.spec.Method != method {
		ctx.AddTag(stringtool.Cat("requestAdaptor: method ", method, " adapted to ", ra.spec.Method))
		req.SetMethod(ra.spec.Method)
	}
```
compared to original
```golang
	if ra.spec.Method != "" && ra.spec.Method != method {
		ctx.AddTag(stringtool.Cat("requestAdaptor: method ", method, " adapted to ", ra.spec.Method))
		req.SetMethod(ra.spec.Method)
	}
```
It become so ugly that I can't bare it. To fix the race condition with support to the all field template, I think it's a good way to place the spec into namespace either. Then it's possible to change the spec as wish as we can. 

I simply removed the template support with Path and Sign field for the `RequestAdaptor` , because it needs some consideration for the race condition.

Reply:
maybe we can merge static spec into template before we call `ra.Builder.Reload(&ra.spec.Spec)` in `reload`, so we can use template to build spec which can be used to modify request/response always in `Handle`

Reply:
the code is ugly in your example above, but it can be not so ugly if we do a little refactoring, like:

```go
newMethod := tempSpec.Method
if newMethod == "" {
    newMethod = ra.spec.Method
}
if newMethod != "" && newMethod != method {
    ctx.AddTag(stringtool.Cat("requestAdaptor: method ", method, " adapted to ", newMethod))
    req.SetMethod(newMethod)
}
```

Reply:
the `sign` field should not be a part of the template, so it is correct to exclude it.
but we **must** include the `path` field, however, we don't need to (and should not) use `pathadaptor.Spec` for it, because in the template, we can easily do more than `pathadaptor.Spec` in a more straightforward way.

Reply:
When you use `newMethod := tempSpec.Method`, the deep-copy exist. So it has the same cost as [fix race conditions](https://github.com/megaease/easegress/pull/910/commits/70070741f70c03298d64afaab12a76b38a2c44aa) does, which using adaptSpecWithTemplate to override the empty value of templateSpec with static spec and has less space cost.


> the code is ugly in your example above, but it can be not so ugly if we do a little refactoring, like:
> 
> ```go
> newMethod := tempSpec.Method
> if newMethod == "" {
>     newMethod = ra.spec.Method
> }
> if newMethod != "" && newMethod != method {
>     ctx.AddTag(stringtool.Cat("requestAdaptor: method ", method, " adapted to ", newMethod))
>     req.SetMethod(newMethod)
> }
> ```

Reply:
It's a good idea. Not `reload` but `build`. I'll try it.

> maybe we can merge static spec into template before we call `ra.Builder.Reload(&ra.spec.Spec)` in `reload`, so we can use template to build spec which can be used to modify request/response always in `Handle`

Reply:
> It's a good idea. Not `reload` but `build`. I'll try it.
> 
> > maybe we can merge static spec into template before we call `ra.Builder.Reload(&ra.spec.Spec)` in `reload`, so we can use template to build spec which can be used to modify request/response always in `Handle`

I withdraw my last opinion. Neither `reload` nor `build` will make it, though it will make code more clean. I think `Reload` only accessed for template changed. `Build` is executed by reference and result in race condition, unless we do a deep copy of static spec.